### PR TITLE
planner: generate index path as candidate when it is a single scan

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

fix https://github.com/pingcap/tidb/issues/12021

When we only want to get primary key, we can consider index path if there is no condition on PK. For example:
```sql
create table t1 (a int primary key, b int, .... many columns, index idx_b(b));
create table t2 (a int, b int);
```

In this query, we can use `IndexScan` for `t1.a`:
```sql
select t2.a, t2.b
from t1, t2
where t1.a = t2.a;
```
We can also use a `IndexScan` for this one:
```sql
select count(1) from t1;
```

More general, we should consider the index path as a candidate when it is a single scan.

### What is changed and how it works?

Change the condition that generates index path.

We can use index to generate physical plan if any of the following conditions is satisfied:
1. This path's access cond is not nil.
2. We have a non-empty prop to match.
3. This index is forced to choose.
4. The needed columns are all covered by index columns(and handleCol).

### Check List 

Tests 

 - Unit test
 - Explain test

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -303,11 +303,13 @@ func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candida
 		var currentCandidate *candidatePath
 		if path.isTablePath {
 			currentCandidate = ds.getTableCandidate(path, prop)
-		} else if len(path.accessConds) > 0 || !prop.IsEmpty() || path.forced {
-			// We will use index to generate physical plan if:
-			// this path's access cond is not nil or
-			// we have prop to match or
-			// this index is forced to choose.
+		} else if len(path.accessConds) > 0 || !prop.IsEmpty() || path.forced ||
+			(len(ds.schema.Columns) == 1 && ds.handleCol != nil && ds.handleCol == ds.schema.Columns[0] && len(path.accessConds) == 0) {
```
Comment (lamxTyler) [inline planner/core/find_best_task.go]: Why not just relax it to single scan?

Comment (francis0407) [reply planner/core/find_best_task.go]: I'm sorry, what does it mean 'relax it to single scan`? Could you please elaborate more on it?

Comment (lamxTyler) [reply planner/core/find_best_task.go]: If index could cover all schema columns, we could also consider it as an candidate.

Comment (francis0407) [reply planner/core/find_best_task.go]: Yeah, it sounds more general. I have fixed that.

**Comment by francis0407** [comment]
Some unit tests about `admin clean up index` and `admin check index` failed. I'm gonna check these tests, maybe fix the issue in another PR.

**Comment by francis0407** [comment]
All the tests have been fixed.
Some of the tests failed before because sql `select count(*) from t` chooses index path, but acutally, they want a table path. If we want to use a table scan, we must use index hint now.

**Pull request merged**